[JIT] Add Support For Module Containers as Iterables

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28255 [JIT] Add Support For Module Containers as Iterables**

Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506

Commit:
Add Support For Module Containers as Iterables

[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And #27790



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And #27790



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1623,56 +1626,40 @@ struct to_ir {
+    if (!iterable->emitUnrolled()) {
+      return emitLoopCommon(loc, body, iterable->getValue(), targets, {});
+    }
+    TORCH_INTERNAL_ASSERT(
```
We should put this logic in a function like emitLoopUnrolled to make it more clear.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1623,56 +1626,40 @@ struct to_ir {
-    pushFrame(environment_stack->block());
-    for (const auto& inst : instances) {
-      emitExprsAssign(targets, {inst}, itrs[0].range(), /*n_binders=*/1);
+    IterableValuePtr iterable = sv->asIterable(loc, method);
```
Suggestion: this can be organized like:
```
// shouldEmitUnrolled throws error if the SV is not iterable at all.
if (sv->shouldEmitUnrolled()) {
  // unrolled
} else {
  // common
}

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -363,8 +393,19 @@ RangeValue::RangeValue(
   }
+
+  // TODO: !has_only_end calculation
+  if (has_only_end_ && toIValue(end_)) {
```
This should be implemented without the TODO or not at all. Not all all is perfectly fine (enumerate can still work by initializing the RangeValue directly with a constant). The intermediate states just bug reports, because we think it works when it really doesn't. This has happened with list comprehensions where since we didn't implement all of the spec the impression is that they do not work at all.

zdevito commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -177,6 +191,53 @@ struct TORCH_API BuiltinFunction : public SugaredValue {
+  };
+
+  std::vector<std::shared_ptr<SugaredValue>> tup_;
+  bool contains_module_list_;
```
This can use more explanation. What are the semantics for when we unroll or do not unroll a loop? Can that be put in a comment somewhere.

zdevito reviewed (commented):
This looks pretty good. I have one substantive comment: that we should try just putting the iteration functionality into sugaredvalue, as a way of avoiding having to introduce a new concept.

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

zdevito commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -14,6 +14,25 @@
 namespace torch {
 namespace jit {
 
+c10::optional<Stack> tryConstantPropNode(const Node* node) {
```
This is missing all of constant props checks that the node itself can be constant-propagated. Doesn't seem safe to use.

zdevito replied:
It seems like this method is actually use as `runNodeIfInputsAreConstant` which has a different implicit contract. Should document what this will do. What is `Print(4)` expected to do here?

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -2507,11 +2504,22 @@ struct to_ir {
+        auto iterable_value = expr_sv->iter(loc, method);
+
+        // range should have the same static length as the other iterable
+        c10::optional<int64_t> iter_static_len = c10::nullopt;
```
Discussed in person: only ranges created from enumerate need to have static lengths. If we do this then we can making have a static length and choosing to unroll the same decision.

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -363,9 +395,26 @@ RangeValue::RangeValue(
+    static_len_ = static_len;
+  } else {
+    auto len_node = len(loc, m)->node();
+    auto maybe_out_stack = tryConstantPropNode(len_node);
```
In person: let's not allow constant prop for ranges unless they are constructed with a static_len

zdevito commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -64,6 +66,12 @@ struct TORCH_API SugaredValue
     throw ErrorReport(loc) << kind() << " cannot be used as a type";
   }
 
+  virtual std::shared_ptr<SugaredValue> iter(
```
can you add a comment for how this gets used and what is expected of the resulting object?

zdevito reviewed (approved):
This looks good. The two substantive comments are: avoid using more constant prop than we were before for ranges, and  lets make statically determined lengths also the same as when we unroll loops to avoid introducing an exponential number of cases to handle.

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Support For Module Containers as Iterables"


Add support for treating Sequentials, ModuleLists, and ModuleDicts as iterables.

As previously, when emitting a for loop over a Module Container we unroll the for loop over all elements. We require that any Sugared Value in an iterable with a Module Container have a statically - determinable length.

Otherwise, if you zipped over a list of varying length and an nn.Sequential that alternated between returning a Tensor and a Dictionary, the output type would change based on the length of the list.

Fix for #17179
And https://github.com/pytorch/pytorch/issues/27401 
and https://github.com/pytorch/pytorch/issues/27506



[ghstack-poisoned]
